pallets/protocol-fee/src/lib.rs

```diff
@@ -0,0 +1,474 @@
+            return Ok(());
+        }
+        let imbalance = Self::withdraw_fee(signatory, fee)?;
+        T::OnProtocolFeePayment::on_unbalanced(imbalance);
+        Ok(())
+    }
+
+    /// Computes the fee for `count` similar operations, and charges that fee to the given
+    /// signatory.
+    pub fn charge_fee_batch(
+        signatory: &Signatory,
+        name: &OperationName,
+        count: usize,
+    ) -> DispatchResult {
+        let fee = Self::compute_fee(name)?.saturating_mul(<BalanceOf<T>>::from(count as u32));
+        let imbalance = Self::withdraw_fee(signatory, fee)?;
+        T::OnProtocolFeePayment::on_unbalanced(imbalance);
+        Ok(())
+    }
+
+    /// Computes the fee of the operation, charges that fee to `signatory`, and pays it out
+    /// collectively to `recipients` in equal parts.
+    pub fn charge_fee_equal_parts(
+        signatory: &Signatory,
+        name: &OperationName,
+        recipients: &[<T as frame_system::Trait>::AccountId],
+    ) -> DispatchResult {
+        let fee = Self::compute_fee(name)?;
+        let mut imbalance = Self::withdraw_fee(signatory, fee)?;
+        let num_recipients = u32::max(1, recipients.len() as u32);
+        let fee_part = imbalance.peek() / num_recipients.into();
+        for account_id in recipients {
+            let (part, rest) = imbalance.split(fee_part);
+            imbalance = rest;
+            T::Currency::resolve_creating(account_id, part);
+        }
+        // Burn the remainder of division by not resolving it.
+        Ok(())
+    }
+
+    /// Withdraws a precomputed fee.
+    fn withdraw_fee(signatory: &Signatory, fee: BalanceOf<T>) -> WithdrawFeeResult<T> {
+        match signatory {
+            Signatory::Identity(did) => T::Currency::withdraw_identity_balance(did, fee)
+                .map_err(|_| Error::<T>::InsufficientIdentityBalance.into()),
+            Signatory::AccountKey(account) => T::Currency::withdraw(
+                &T::AccountId::decode(&mut &account.encode()[..])
+                    .map_err(|_| Error::<T>::AccountIdDecode)?,
+                fee,
+                WithdrawReason::Fee.into(),
+                ExistenceRequirement::KeepAlive,
+            )
+            .map_err(|_| Error::<T>::InsufficientAccountBalance.into()),
+        }
+    }
+}
+
+impl<T: Trait> ChargeProtocolFee<T::AccountId> for Module<T> {
+    fn charge_fee(signatory: &Signatory, name: &OperationName) -> DispatchResult {
+        Self::charge_fee(signatory, name)
+    }
+
+    fn charge_fee_batch(
+        signatory: &Signatory,
+        name: &OperationName,
+        count: usize,
+    ) -> DispatchResult {
+        Self::charge_fee_batch(signatory, name, count)
+    }
+
+    fn charge_fee_equal_parts(
+        signatory: &Signatory,
+        name: &OperationName,
+        recipients: &[T::AccountId],
+    ) -> DispatchResult {
+        Self::charge_fee_equal_parts(signatory, name, recipients)
+    }
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
To make maintenance of tests easier, can you please use the storage in `pallets/runtime/src/test/storage.rs`. You might have to move the test cases to the runtime.

Reply:
I understand the maintenance concern and moving the tests is not a problem. However it would have been better to have test storage outside the runtime crate to allow tests from other modules like `protocol_fee` which don't depend on the runtime crate to be implemented outside the runtime crate.